var/spack/repos/builtin/packages/py-torch-nvidia-apex/package.py

```diff
@@ -13,44 +13,164 @@ class PyTorchNvidiaApex(PythonPackage, CudaPackage):
+    variant(
+        "fast_multihead_attn", default=False, description="Build fast multihead attention module"
+    )
+    variant("transducer", default=False, description="Build transducer module")
+    variant("cudnn_gbn_lib", default=False, description="Build cudnn gbn module")
+    variant("peer_memory_cuda", default=False, description="Build peer memory module")
+    variant("nccl_p2p_cuda", default=False, description="Build with nccl p2p")
+    variant("fast_bottleneck", default=False, description="Build fast_bottleneck module")
+    variant("fused_conv_bias_relu", default=False, description="Build fused_conv_bias_relu moduel")
+
+    requires(
+        "+peer_memory_cuda+nccl_p2p_cuda",
+        when="+fast_bottleneck",
+        msg="+fast_bottleneck requires both +peer_memory_cuda and +nccl_p2p_cuda to be enabled.",
+    )
+    requires("^cudnn@8.5:", when="+cudnn_gbn_lib")
+    requires("^cudnn@8.4:", when="+fused_conv_bias_relu")
+    requires("^nccl@2.10:", when="+nccl_p2p_cuda")
+
+    with default_args(type=("build")):
+        depends_on("py-setuptools")
+        depends_on("py-packaging")
+        depends_on("py-pip")
+    with default_args(type=("build", "run")):
+        depends_on("python@3:")
+        depends_on("py-torch@0.4:")
+        for _arch in CudaPackage.cuda_arch_values:
+            depends_on(f"py-torch+cuda cuda_arch={_arch}", when=f"+cuda cuda_arch={_arch}")
+
+    depends_on("py-pybind11", type=("build", "link", "run"))
+    depends_on("cuda@9:", when="+cuda")
+
     # https://github.com/NVIDIA/apex/issues/1498
     # https://github.com/NVIDIA/apex/pull/1499
     patch("1499.patch", when="@2020-10-19")
 
+    conflicts(
+        "cuda_arch=none",
+        when="+cuda",
+        msg="Must specify CUDA compute capabilities of your GPU, see "
+        "https://developer.nvidia.com/cuda-gpus",
+    )
+
+    def torch_cuda_arch_list(self, env):
+        if self.spec.satisfies("+cuda"):
+            torch_cuda_arch = ";".join(
+                "{0:.1f}".format(float(i) / 10.0) for i in self.spec.variants["cuda_arch"].value
+            )
+            env.set("TORCH_CUDA_ARCH_LIST", torch_cuda_arch)
+
     def setup_build_environment(self, env):
-        if "+cuda" in self.spec:
+        if self.spec.satisfies("+cuda"):
             env.set("CUDA_HOME", self.spec["cuda"].prefix)
+            self.torch_cuda_arch_list(env)
         else:
             env.unset("CUDA_HOME")
 
-    @when("^python@:3.10")
+    def setup_run_environment(self, env):
+        self.torch_cuda_arch_list(env)
+
+    @when("^py-pip@:23.0")
     def global_options(self, spec, prefix):
         args = []
         if spec.satisfies("^py-torch@1.0:"):
             args.append("--cpp_ext")
-            if "+cuda" in spec:
+            if spec.satisfies("+cuda"):
                 args.append("--cuda_ext")
+
+        if spec.satisfies("+permutation_search_cuda"):
+            args.append("--permutation_search")
```

Comment:
Optional.

Maybe a little helper lambda would be useful here:
```suggestion
        variant_to_arg = lambda v: args.append(f"--{v}") if spec.satisfies(f"+{v}") else None
        variant_to_arg("permutation_search_cuda")
        # etc.
```
Requires that the variant names match the options exactly, unless you want a map between variants and options? Keeping the same names for both might be more straightforward, but you may also have good reasons for adjusting the names, not sure.

Same comment applies to `config_settings`.

Reply:
Yes I thought about that but I don't know why in the README of apex they don't have matching names between name and options. I think i could just decide to use the naming of the setup.py options i.e. the one with `--`. I will update this accordingly

Reply:
As this is an optional comment and there was no other comment, I approve.